drivers/dma/dma_intel_adsp_hda.c

```diff
@@ -349,6 +353,42 @@ int intel_adsp_hda_dma_stop(const struct device *dev, uint32_t channel)
 		return 0;
 	}
 
+	/*
+	 * Workaround for random metallic noise on DMA restart:
+	 * Make sure that the output link DMA read and write pointers are exactly 0 when we disable
+	 * the channel.
+	 * On channel restart the pointers are not rested to 0 and can result corruption.
+	 */
+	if (cfg->direction == MEMORY_TO_PERIPHERAL) {
+		uint32_t wp, size;
+		int i;
+
+		wp = *DGBWP(cfg->base, cfg->regblock_size, channel);
+		size = intel_adsp_hda_get_buffer_size(cfg->base, cfg->regblock_size, channel);
+
+		/* Move the WP to wrap to exactly to 0 if it is not already 0 */
+		if (wp) {
+			intel_adsp_hda_link_commit(cfg->base, cfg->regblock_size, channel,
+						   size - wp);
+		}
+
+		/* Wait for the pointers to reach 0 */
+		for (i = 0; i < 1000; i++) {
+			uint32_t unused = intel_adsp_hda_unused(cfg->base, cfg->regblock_size,
+								channel);
+
+			if (unused >= size)
+				break;
+			k_sleep(K_USEC(1));
```

Comment:
k_sleep(K_USEC(1)) is pretty unlikely to sleep for anything less than 100 microseconds or so, is this what you wanted? If not then there's other ways to loop, yielding control, for a duration of time.